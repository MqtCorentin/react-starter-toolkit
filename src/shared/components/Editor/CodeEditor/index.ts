export { default, useCodeEditor } from './CodeEditor';
