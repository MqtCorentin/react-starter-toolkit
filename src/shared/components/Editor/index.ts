export { withEditor, useEditable } from './Editor';
export type { TEvent, Tknobs } from './Editor';
