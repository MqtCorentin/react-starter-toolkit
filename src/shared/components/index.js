export { default as ModalCommonBody } from './ModalCommonBody';
export { default as ModalCommonHeader } from './ModalCommonHeader';
export { default as ModalCommonFooter } from './ModalCommonFooter';
