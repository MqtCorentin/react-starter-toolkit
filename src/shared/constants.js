import { string, number, shape, oneOfType, arrayOf } from 'prop-types';

export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_FORMAT = 'Format invalide';

export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const rulesRequired = {
  required: {
    message: MSG_REQUIRED,
  },
};

export const rulePattern = {
  pattern: {
    regex: /^[0-9a-zA-Z\s-_]{1,50}$/,
    message: MSG_FORMAT,
  },
};

export const rulesErrorFormat = {
  required: {
    message: MSG_ERROR_FORMAT,
  },
};

export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

export const TIMEOUT = 20000;
export const TIMEOUT_ERROR_MESSAGE = 'Erreur: la requête a mis trop de temps';

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

export const COMMENT = {
  NAME: 'comment',
  LABEL: 'Commentaire',
  SUBLABEL: 'Facultatif',
  MAXLENGTH: '300',
};

export const JSON_CONTENT_HEADER = {
  'Content-Type': 'application/json',
};

export const STORYBOOK = 'https://axaguildev.github.io/react-toolkit/latest/storybook/';
export const DESIGN_SYSTEM = 'https://axaguildev.github.io/design-system/';

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';

export const PROFILS = [ADMIN, READER, WRITER, USER];

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

/**
 * Commons proptypes
 */

export const ContextPropTypes = shape({
  Provider: shape({ defaultValue: string, currentValue: string, changedBits: number }),
  Consumer: shape({ defaultValue: string, currentValue: string, changedBits: number }),
});

export const SelectPropTypes = arrayOf(
  shape({
    value: oneOfType([string, number]),
    label: string,
  }),
);

export const AnomalyPropTypes = shape({
  code: number,
  detail: string,
  label: string,
});

export const IdPropTypes = oneOfType([string, number]);
