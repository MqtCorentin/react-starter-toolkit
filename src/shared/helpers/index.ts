export { formatDate, isValidDate, setDate } from './formatDate';
export { default as setInitialState } from './setInitialState';
export { default as isEmptyOrNull } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setResponseError } from './setResponseError';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as setLoaderMode } from './setLoaderMode';
export { genericHandleChange, getErrorsList, computeInitialStateErrorMessage, getValuesList } from './validation.generic';
export { setTextTooltip, setDisplay } from './formatDataTable';
