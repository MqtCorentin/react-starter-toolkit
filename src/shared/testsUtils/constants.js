export const MOCK_API_URL = 'https://react-starter-api.vercel.app/api/';
