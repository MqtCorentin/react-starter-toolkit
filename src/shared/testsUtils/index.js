export { default as emptyFunction } from './emptyFunction';
export { default as WrapperStaticRouter } from './WrapperStaticRouter';
export { default as renderWithWrapperStaticRouter } from './renderWithWrapperStaticRouter';
