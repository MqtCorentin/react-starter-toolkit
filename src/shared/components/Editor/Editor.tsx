import { ComponentType, useCallback, useState } from 'react';
import { Text, CheckboxItem, CheckboxModes, Select } from '@axa-fr/react-toolkit-all';
import Draggable from 'react-draggable';
import Icons from 'shared/components/Icons';
import { ClickEvent } from '@axa-fr/react-toolkit-core';
import { DEFAULT_OPTION_LABEL, DESIGN_SYSTEM, GITHUB, STORYBOOK } from 'shared/constants';
import './Editor.scss';

export type TEvent = {
  value: string;
  id?: string;
  name?: string;
};

const omittedProps = ['onChange', 'knobs', 'onClick'];

type TListSelect = {
  value: string | number;
  options: Record<string, string>;
};

type TFieldEditor = {
  value: string | boolean | TListSelect;
  name: string;
  onChange: (e: TEvent) => void;
  options?: { label: string; value: string }[];
};

export type Tknobs = Record<string, Record<string, unknown>>;

export const FieldEditor = ({ value, name, onChange }: TFieldEditor) => (
  <>
    {(() => {
      switch (true) {
        case typeof value === 'boolean':
          return (
            <CheckboxItem
              disabled={false}
              isChecked={value}
              onChange={onChange}
              id={name}
              name={name}
              value={`${value}`}
              mode={CheckboxModes.toggle}
              className="af-form__checkbox-toggle"
            />
          );
        case typeof value === 'object':
          return (
            typeof value === 'object' && (
              <div className="af-form__select">
                <Select
                  forceDisplayPlaceholder={true}
                  placeholder={DEFAULT_OPTION_LABEL}
                  name={name}
                  id={name}
                  options={value.options}
                  onChange={onChange}
                  value={value.value}
                  aria-label={`select-${name}`}
                />
              </div>
            )
          );

        default:
          return <Text type="text" value={value} id={name} name={name} onChange={onChange} />;
      }
    })()}
  </>
);

type TmergePropsAndKnobs<P> = {
  props: P & Record<string, unknown>;
  knobs: Tknobs;
};

export const mergePropsAndKnobs = <P extends object>({ props, knobs }: TmergePropsAndKnobs<P>) =>
  Object.keys(props).reduce(
    (acc, key) => ({
      ...acc,
      [key]: knobs[key] ? { ...knobs[key], value: props[key] } : props[key],
    }),
    {},
  );

/**
 * @param onChange : function for update state fields value
 * @param knobs : overrided props (for object type)
 * @returns Form fields editor
 */
export const FormEditor = <P extends object>({
  onChange,
  knobs,
  ...props
}: Omit<P, 'onClick'> & { onChange: ReturnType<typeof useEditable>['onChange']; knobs: Tknobs }) => (
  <form className="af-form-editor">
    <>
      {Object.entries(mergePropsAndKnobs({ props, knobs }))
        .filter(([key]) => !omittedProps.includes(key))
        .map(([name, value]) => (
          <div className="af-form-editor__field" key={name}>
            <label className="af-form__group-label" htmlFor={name}>
              {name}
            </label>
            <FieldEditor key={name} name={name} value={value as TFieldEditor['value']} onChange={onChange(name)} />
          </div>
        ))}
    </>
  </form>
);

export const useToggleEditor = (initState = false) => {
  const [isOpenEditor, setStateToggleEditor] = useState(initState);

  const closeEditor = useCallback(() => {
    setStateToggleEditor(false);
  }, []);

  const openEditor = useCallback(() => {
    setStateToggleEditor(true);
  }, []);

  return { closeEditor, openEditor, isOpenEditor };
};

export type TReturnUseToggleEditor = ReturnType<typeof useToggleEditor>;

/**
 * HOC to add Editor
 * @param Component : component to be wrapped
 * @param knobs : for lists props values
 * @returns Component Editable
 */
export const withEditor =
  <P extends object>(
    Component: ComponentType<P>,
    knobs: Tknobs = {},
    className = 'af-editor',
    useToggleEditorFn = useToggleEditor,
  ): ComponentType<P> =>
  (props: P) => {
    const { closeEditor, openEditor, isOpenEditor } = useToggleEditorFn();
    return (
      <section className={className}>
        <Component {...props} openEditor={openEditor} />
        <div className={`af-draggable-container${isOpenEditor ? ' af-draggable-container--open' : ''}`}>
          <Draggable handle="strong">
            <div className="af-draggable">
              <h3 className="af-draggable__title">
                Props Editor
                <div className="af-draggable__tools">
                  <strong className="glyphicon glyphicon-move"></strong>
                  <i className="glyphicon glyphicon-close no-cursor" onClick={closeEditor}></i>
                </div>
              </h3>

              <div className="af-draggable__content">
                <FormEditor<P> {...(props as P & { onChange: ReturnType<typeof useEditable>['onChange'] })} knobs={knobs} />
              </div>
            </div>
          </Draggable>
        </div>
      </section>
    );
  };

export const setValue = (value: string) => {
  if (value === 'false') return true;
  if (value === 'true') return false;

  return value;
};

type TuseEditable<T> = {
  initialState: T;
  logEventFn?: typeof console.log;
  setValueFn?: typeof setValue;
};
export const useEditable = <T extends object>({ initialState, logEventFn = console.log, setValueFn = setValue }: TuseEditable<T>) => {
  const [state, setState] = useState(initialState);

  const onClick = useCallback(
    (e: ClickEvent) => {
      logEventFn('click button event', e);
    },
    [logEventFn],
  );

  const onChange = useCallback(
    (key: string) => (e: TEvent) => {
      setState(prevState => ({ ...prevState, [key]: setValueFn(e.value) }));
    },
    [setValueFn],
  );
  return { onClick, onChange, state };
};

type TEditorHeader = {
  storybookPath?: string;
  designSystemPath?: string;
  githubPackage?: string;
} & Partial<TReturnUseToggleEditor>;

export const EditorHeader = ({ storybookPath = '', designSystemPath = '', githubPackage = '', openEditor }: TEditorHeader) => (
  <header className="af-editor__header">
    {!!designSystemPath && (
      <a className="af-link" href={`${DESIGN_SYSTEM}${designSystemPath}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="slash" viewBox="0 0 342.988 280" />
        Guidelines
      </a>
    )}

    {!!storybookPath && (
      <a className="af-link" href={`${STORYBOOK}${storybookPath}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="storybook" viewBox="0 0 9.6 12" />
        Storybook
      </a>
    )}

    {!!githubPackage && (
      <a className="af-link" href={`${GITHUB}${githubPackage}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="github" viewBox="0 0 438 438" />
        Github
      </a>
    )}

    <span className="af-link" onClick={openEditor}>
      <i className="glyphicon glyphicon-cog"></i>
      <span className="af-link">Edit props</span>
    </span>
  </header>
);
