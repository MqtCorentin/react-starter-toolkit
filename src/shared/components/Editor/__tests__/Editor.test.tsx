import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { Text } from '@axa-fr/react-toolkit-all';
import { emptyFunction } from 'shared/testsUtils';
import { FieldEditor, mergePropsAndKnobs, withEditor, setValue, useEditable, useToggleEditor, TEvent } from '../Editor';

describe('setValue', () => {
  it.each`
    value      | expected
    ${'true'}  | ${false}
    ${'false'} | ${true}
    ${'value'} | ${'value'}
  `('Should return expected: $expected when value: $value', ({ value, expected }) => {
    const result = setValue(value);
    expect(result).toEqual(expected);
  });
});

describe('mergePropsAndKnobs', () => {
  const otherProps = {
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
  };
  const props = {
    ...otherProps,
    classModifier: 'error',
  };
  const knobs = {
    classModifier: {
      value: '',
      options: [
        {
          label: 'Success',
          value: 'success',
        },
      ],
    },
  };
  const expectedMerge = { ...otherProps, classModifier: { ...knobs.classModifier, value: 'error' } };

  it.each`
    props    | knobs    | expected
    ${{}}    | ${{}}    | ${{}}
    ${props} | ${knobs} | ${expectedMerge}
  `('Should return expected: $expected when props: $props, knobs: $knobs', ({ props, knobs, expected }) => {
    const result = mergePropsAndKnobs({ props, knobs });
    expect(result).toEqual(expected);
  });
});

describe('FieldEditor', () => {
  const onChange = jest.fn();
  it.each`
    value
    ${true}
    ${'value'}
    ${{ value: 'value', options: [] }}
    ${{ value: 'value', type: 'jsx', labelBtnOpenCodeEditor: 'Edit JSX' }}
    ${{ value: 'value', type: 'json', labelBtnOpenCodeEditor: 'Edit Json' }}
    ${emptyFunction}
  `('Should render when value: $value', ({ value }) => {
    const { asFragment } = render(<FieldEditor value={value} name="name" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useEditable', () => {
  const initialState = {
    classModifier: '',
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
    autoFocus: false,
  };

  it('Should logEventFn have been called when onClick have been called', () => {
    const logEventFn = jest.fn();
    const { result } = renderHook(() => useEditable({ initialState, logEventFn }));
    act(() => result.current.onClick('name')({ id: 'id' }));
    expect(logEventFn).toHaveBeenCalledWith('click event', 'name', { id: 'id' });
  });

  it('Should setValueFn have been called and return updated state when onChange have been called with event', () => {
    const setValueFn = jest.fn().mockReturnValue('setted value');
    const { result } = renderHook(() => useEditable({ initialState, setValueFn }));
    act(() => result.current.onChange('classModifier')({ value: 'newvalue' }));
    expect(setValueFn).toHaveBeenCalledWith('newvalue');
    expect(result.current.state.classModifier).toEqual('setted value');
  });

  it('Should set autofocus state to false when onBlur have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: true } }));
    act(() => result.current.onBlur());
    expect(result.current.state.autoFocus).toBeFalsy();
  });

  it('Should set autofocus state to true when onBlur have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: false } }));
    act(() => result.current.onFocus());
    expect(result.current.state.autoFocus).toBeTruthy();
  });
});

describe('withEditor', () => {
  type Props = { className?: string; onChange: ReturnType<typeof useEditable>['onChange'] };
  const onChange = jest.fn().mockImplementation(name => (e: TEvent) => {
    console.log('onChange');
  });
  const Component = ({ className, onChange }: Props) => (
    <Text id="idtext" name="nametext" type="text" value="hello" onChange={onChange} className={className} />
  );

  it('Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true', () => {
    const useToggleEditorFn = jest.fn().mockReturnValue({
      closeEditor: jest.fn(),
      openEditor: jest.fn(),
      isOpenEditor: true,
    });
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFn);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render Component with FormEditor when apply withEditor HOC', () => {
    const ComponentWithEditor = withEditor<Props>(Component);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const useToggleEditorFnNoOpen = jest.fn().mockReturnValue({
    closeEditor: jest.fn(),
    openEditor: jest.fn(),
    isOpenEditor: false,
  });
  it("Shouldn't render Component with isOpenEditor false", () => {
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFnNoOpen);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useToggleEditor', () => {
  it('Should isOpenEditor to be false when closeEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.closeEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(false);
  });

  it('Should isOpenEditor to be true when openEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.openEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(true);
  });
});
