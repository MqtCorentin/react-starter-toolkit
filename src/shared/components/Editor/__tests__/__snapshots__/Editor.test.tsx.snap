// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldEditor Should render when value: [Function emptyFuncion] 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-text"
    disabled=""
    id="name"
    name="name"
    readonly=""
    type="text"
    value="() => null"
  />
</DocumentFragment>
`;

exports[`FieldEditor Should render when value: {"labelBtnOpenCodeEditor": "Edit JSX", "type": "jsx", "value": "value"} 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit JSX
    </span>
  </button>
</DocumentFragment>
`;

exports[`FieldEditor Should render when value: {"labelBtnOpenCodeEditor": "Edit Json", "type": "json", "value": "value"} 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit Json
    </span>
  </button>
</DocumentFragment>
`;

exports[`FieldEditor Should render when value: {"options": [Array], "value": "value"} 1`] = `
<DocumentFragment>
  <div
    class="af-form__select"
  >
    <div
      class="af-form__select-container"
    >
      <select
        aria-label="select-name"
        class="af-form__input-select"
        id="name"
        name="name"
      />
      <span
        aria-controls="name"
        class="glyphicon glyphicon-menu-down"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldEditor Should render when value: true 1`] = `
<DocumentFragment>
  <div
    class="af-form__checkbox-toggle"
  >
    <input
      checked=""
      class="af-form__input-checkbox"
      id="name"
      mode="toggle"
      name="name"
      type="checkbox"
    />
    <label
      class="af-form__label"
      for="name"
    >
      <span
        class="af-form__indicator"
      >
        <i
          class="glyphicon glyphicon-ok"
        />
      </span>
      <span
        class="af-form__description"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`FieldEditor Should render when value: value 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-text"
    id="name"
    name="name"
    type="text"
    value="value"
  />
</DocumentFragment>
`;

exports[`withEditor Should render Component with FormEditor when apply withEditor HOC 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;

exports[`withEditor Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
    <div
      class="af-draggable-container react-draggable"
      style="transform: translate(0px,0px);"
    >
      <div
        class="af-draggable"
      >
        <h3
          class="af-draggable__title"
        >
          Props Editor
          <div
            class="af-draggable__tools"
          >
            <i
              class="glyphicon glyphicon-move"
            />
            <i
              class="glyphicon glyphicon-close"
            />
          </div>
        </h3>
        <div
          class="af-draggable__content"
        >
          <form
            class="af-form-editor"
          >
            <div
              class="af-form-editor__field"
            >
              <label
                class="af-form__group-label"
                for="className"
              >
                className
              </label>
              <input
                class="af-form__input-text"
                id="className"
                name="className"
                type="text"
                value="af-component"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`withEditor Shouldn't render Component with isOpenEditor false 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;
